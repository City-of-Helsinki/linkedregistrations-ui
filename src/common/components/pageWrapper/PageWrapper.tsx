import classNames from 'classnames';
import React from 'react';

import upperCaseFirstLetter from '../../../utils/upperCaseFirstLetter';

import styles from './pageWrapper.module.scss';

export interface PageWrapperProps {
  backgroundColor?: 'white' | 'gray' | 'coatOfArms';
  className?: string;
  /**
   * Translation path of the page title. Translation will be overriden by titleText if defined
   */
}

const PageWrapper: React.FC<React.PropsWithChildren<PageWrapperProps>> = ({
  backgroundColor = 'white',
  children,
  className,
}) => {
  return (
    <div
      className={classNames(
        styles.pageWrapper,
        [styles[`backgroundColor${upperCaseFirstLetter(backgroundColor)}`]],
        className
      )}
    >
      {children}
    </div>
  );
};

export default PageWrapper;
