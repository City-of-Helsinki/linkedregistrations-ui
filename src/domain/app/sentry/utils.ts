import * as Sentry from '@sentry/browser';

const reportError = ({
  data,
  message,
}: {
  data: { error: Record<string, unknown> } & Record<string, unknown>;
  message: string;
}): string => {
  const { error, ...restData } = data;
  const reportObject = {
    extra: {
      data: {
        ...restData,
        currentUrl: window.location.href,
        timestamp: new Date(),
        userAgent: navigator.userAgent,
        errorAsString: JSON.stringify(error),
      },
      level: 'error',
    },
  };

  return Sentry.captureException(message, reportObject);
};

export { reportError };
