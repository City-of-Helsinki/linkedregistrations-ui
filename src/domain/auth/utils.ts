/* eslint-disable no-console */
import axios from 'axios';

import { APITokens, ExtendedSession, RefreshTokenResponse } from '../../types';
import { User } from '../user/types';

export const getApiTokensRequest = async ({
  accessToken,
  linkedEventsApiScope,
  url,
}: {
  accessToken: string;
  linkedEventsApiScope: string;
  url: string;
}): Promise<APITokens> => {
  const response = await axios.post(url, undefined, {
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    responseType: 'json',
  });

  const linkedevents = response.data[linkedEventsApiScope];

  return { linkedevents };
};

export const refreshAccessTokenRequest = async ({
  clientId,
  clientSecret,
  refreshToken,
  url,
}: {
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  url: string;
}): Promise<RefreshTokenResponse> => {
  const response = await axios.post(
    url,
    {
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
    },
    {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      responseType: 'json',
    }
  );

  return response.data;
};

export const getUserName = ({
  session,
  user,
}: {
  session: ExtendedSession | null;
  user: User | undefined;
}): string => user?.display_name || session?.user?.email || '';
