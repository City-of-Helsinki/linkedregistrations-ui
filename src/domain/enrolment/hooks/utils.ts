import { TFunction } from 'i18next';

import { LEServerError, ServerErrorItem } from '../../../types';
import parseServerErrorMessage from '../../../utils/parseServerErrorMessage';
import pascalCase from '../../../utils/pascalCase';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ErrorObject = Record<string, any>;
type ErrorType = ErrorObject | ErrorObject[] | string;

export const parseEnrolmentServerErrors = ({
  error,
  t,
}: {
  error: ErrorType;
  t: TFunction;
}): ServerErrorItem[] => {
  // LE returns errors as array when trying to create/edit multiple enrolments in same request.
  // In that case call parseEnrolmentServerErrors recursively to get all single errors
  if (Array.isArray(error)) {
    return (error as ErrorObject[]).reduce(
      (previous: ServerErrorItem[], r: ErrorType) => [
        ...previous,
        ...parseEnrolmentServerErrors({ error: r, t }),
      ],
      []
    );
  }

  return typeof error === 'string'
    ? [{ label: '', message: parseServerErrorMessage({ error: [error], t }) }]
    : Object.entries(error).reduce(
        (previous: ServerErrorItem[], [key, error]) => [
          ...previous,
          ...parseEnrolmentServerError({ error, key }),
        ],
        []
      );

  // Get error item for an single error.
  function parseEnrolmentServerError({
    error,
    key,
  }: {
    error: LEServerError;
    key: string;
  }) {
    switch (key) {
      case 'detail':
      case 'non_field_errors':
        return [{ label: '', message: parseServerErrorMessage({ error, t }) }];
      default:
        return [
          {
            label: parseEnrolmentServerErrorLabel({ key }),
            message: parseServerErrorMessage({ error, t }),
          },
        ];
    }
  }

  // Get correct field name for an error item
  function parseEnrolmentServerErrorLabel({ key }: { key: string }): string {
    return t(`enrolment:label${pascalCase(key)}`);
  }
};
