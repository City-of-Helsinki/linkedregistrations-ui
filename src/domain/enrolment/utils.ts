import { AxiosError } from 'axios';

import formatDate from '../../utils/formatDate';
import stringToDate from '../../utils/stringToDate';
import axiosClient from '../app/axios/axiosClient';
import { Registration } from '../registration/types';
import {
  ENROLMENT_INITIAL_VALUES,
  NOTIFICATIONS,
  NOTIFICATION_TYPE,
} from './constants';
import {
  CreateEnrolmentMutationInput,
  Enrolment,
  EnrolmentFormFields,
} from './types';

export const createEnrolment = async (
  input: CreateEnrolmentMutationInput
): Promise<Enrolment> => {
  try {
    const { data } = await axiosClient.post('/signup/', JSON.stringify(input));
    return data;
  } catch (error) {
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const deleteEnrolment = async (
  cancellationCode: string
): Promise<null> => {
  try {
    const { data } = await axiosClient.delete('/signup/', {
      data: JSON.stringify({ cancellation_code: cancellationCode }),
    });
    return data;
  } catch (error) {
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const getEnrolmentNotificationsCode = (
  notifications: string[]
): NOTIFICATION_TYPE => {
  if (
    notifications.includes(NOTIFICATIONS.EMAIL) &&
    notifications.includes(NOTIFICATIONS.SMS)
  ) {
    return NOTIFICATION_TYPE.SMS_EMAIL;
  } else if (notifications.includes(NOTIFICATIONS.EMAIL)) {
    return NOTIFICATION_TYPE.EMAIL;
  } else if (notifications.includes(NOTIFICATIONS.SMS)) {
    return NOTIFICATION_TYPE.SMS;
  } else {
    return NOTIFICATION_TYPE.NO_NOTIFICATION;
  }
};

export const getEnrolmentNotificationTypes = (
  notifications: string
): NOTIFICATIONS[] => {
  switch (notifications) {
    case NOTIFICATION_TYPE.SMS:
      return [NOTIFICATIONS.SMS];
    case NOTIFICATION_TYPE.EMAIL:
      return [NOTIFICATIONS.EMAIL];
    case NOTIFICATION_TYPE.SMS_EMAIL:
      return [NOTIFICATIONS.EMAIL, NOTIFICATIONS.SMS];
    default:
      return [];
  }
};

export const getEnrolmentPayload = (
  formValues: EnrolmentFormFields,
  registration: Registration
): CreateEnrolmentMutationInput => {
  const {
    city,
    dateOfBirth,
    email,
    extraInfo,
    membershipNumber,
    name,
    nativeLanguage,
    notifications,
    phoneNumber,
    serviceLanguage,
    streetAddress,
    zip,
  } = formValues;

  return {
    city: city || null,
    date_of_birth: dateOfBirth
      ? formatDate(stringToDate(dateOfBirth), 'yyyy-MM-dd')
      : null,
    email: email || null,
    extra_info: extraInfo,
    membership_number: membershipNumber,
    name: name || null,
    native_language: nativeLanguage || null,
    notifications: getEnrolmentNotificationsCode(notifications),
    phone_number: phoneNumber || null,
    registration: registration.id as string,
    service_language: serviceLanguage || null,
    street_address: streetAddress || null,
    zipcode: zip || null,
  };
};

export const getEnrolmentDefaultInitialValues = (
  registration: Registration
): EnrolmentFormFields => ({
  ...ENROLMENT_INITIAL_VALUES,
  audienceMaxAge: registration.audience_max_age ?? null,
  audienceMinAge: registration.audience_min_age ?? null,
});

export const getEnrolmentInitialValues = (
  enrolment: Enrolment,
  registration: Registration
): EnrolmentFormFields => {
  return {
    ...getEnrolmentDefaultInitialValues(registration),
    accepted: true,
    city: enrolment.city ?? '-',
    dateOfBirth: enrolment.date_of_birth
      ? formatDate(new Date(enrolment.date_of_birth))
      : '',
    email: enrolment.email ?? '-',
    extraInfo: enrolment.extra_info ?? '-',
    membershipNumber: enrolment.membership_number ?? '-',
    name: enrolment.name ?? '-',
    nativeLanguage: enrolment.native_language ?? '',
    notifications: getEnrolmentNotificationTypes(
      enrolment.notifications as string
    ),
    phoneNumber: enrolment.phone_number ?? '-',
    serviceLanguage: enrolment.service_language ?? '',
    streetAddress: enrolment.street_address ?? '-',
    zip: enrolment.zipcode ?? '-',
  };
};
