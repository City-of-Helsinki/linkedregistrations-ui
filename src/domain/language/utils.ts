import { AxiosError } from 'axios';
import capitalize from 'lodash/capitalize';
import get from 'lodash/get';

import { OptionType, Language, ExtendedSession } from '../../types';
import getLocalisedString from '../../utils/getLocalisedString';
import queryBuilder from '../../utils/queryBuilder';
import { callGet } from '../app/axios/axiosClient';

import {
  LanguagesQueryVariables,
  LanguagesResponse,
  LELanguage,
} from './types';

export const fetchLanguages = async (
  args: LanguagesQueryVariables,
  session: ExtendedSession | null
): Promise<LanguagesResponse> => {
  try {
    const { data } = await callGet({
      session,
      url: languagesPathBuilder(args),
    });
    return data;
  } catch (error) {
    /* istanbul ignore next */
    throw Error(JSON.stringify((error as AxiosError).response?.data));
  }
};

export const languagesPathBuilder = (args: LanguagesQueryVariables): string => {
  const { serviceLanguage } = args;
  const variableToKeyItems = [
    { key: 'service_language', value: serviceLanguage },
  ];

  const query = queryBuilder(variableToKeyItems);

  return `/language/${query}`;
};

export const getLanguageOption = (
  language: LELanguage,
  locale: Language
): OptionType => ({
  label: capitalize(getLocalisedString(language.name, locale)),
  value: language.id,
});

export const sortLanguageOptions = (a: OptionType, b: OptionType): number => {
  const languagePriorities = {
    fi: 3,
    sv: 2,
    en: 1,
  };
  const aPriority = get(languagePriorities, a.value, 0);
  const bPriority = get(languagePriorities, b.value, 0);

  if (aPriority !== bPriority) {
    return bPriority - aPriority;
  }
  return a.label.localeCompare(b.label, undefined, { sensitivity: 'base' });
};
