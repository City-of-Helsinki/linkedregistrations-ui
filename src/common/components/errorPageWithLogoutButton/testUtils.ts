import * as nextAuth from 'next-auth/react';

import { screen, userEvent } from '../../../utils/testUtils';

const getSignOutButton = () => {
  return screen.getByRole('button', { name: /kirjaudu ulos/i });
};

export const shouldRenderErrorPageTexts = async ({
  text,
  title,
}: {
  text: string;
  title: string;
}) => {
  expect(screen.getByRole('heading', { name: title })).toBeInTheDocument();
  expect(screen.getByText(text)).toBeInTheDocument();
};

export const shouldSignOut = async () => {
  const user = userEvent.setup();
  jest
    .spyOn(nextAuth, 'signOut')
    .mockImplementation()
    .mockResolvedValue({ url: 'https://test.com' });

  const signoutButton = getSignOutButton();
  await user.click(signoutButton);

  expect(nextAuth.signOut).toBeCalled();
};
