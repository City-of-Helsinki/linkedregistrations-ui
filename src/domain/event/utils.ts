import { DateArray, DateTime, EventAttributes, createEvent } from 'ics';
import { TFunction } from 'next-i18next';

import { AddNotificationFn } from '../../common/components/notificationsContext/NotificationsContext';
import { Language } from '../../types';
import getLocalisedString from '../../utils/getLocalisedString';
import { getPlaceFields } from '../place/utils';

import { Event, EventFields } from './types';

export const getEventFields = (event: Event, locale: Language): EventFields => {
  return {
    audienceMaxAge: event.audience_max_age || null,
    audienceMinAge: event.audience_min_age || null,
    description: getLocalisedString(event.description, locale),
    endTime: event.end_time ? new Date(event.end_time) : null,
    freeEvent: !!event.offers[0]?.is_free,
    imageUrl: event.images.find((image) => image?.url)?.url || null,
    keywords: event.keywords.filter((id) => id),
    name: getLocalisedString(event.name, locale),
    offers: event.offers.filter((offer) => !!offer && !offer?.is_free),
    shortDescription: getLocalisedString(event.short_description, locale),
    startTime: event.start_time ? new Date(event.start_time) : null,
  };
};

export const getEventLocationText = ({
  event,
  locale,
}: {
  event: Event;
  locale: Language;
}): string => {
  const { location } = event;
  const {
    addressLocality,
    name: locationName,
    streetAddress,
  } = location
    ? getPlaceFields(location, locale)
    : { addressLocality: '', name: '', streetAddress: '' };

  return (
    [locationName, streetAddress, addressLocality]
      .filter((e) => e)
      .join(', ') || '-'
  );
};

const getDateArray = (date: Date): DateArray => [
  date.getFullYear(),
  date.getMonth() + 1,
  date.getDate(),
  date.getHours(),
  date.getMinutes(),
];

export const getEventAttributes = ({
  event,
  locale,
}: {
  event: Event;
  locale: Language;
}) => {
  const { name, shortDescription } = getEventFields(event, locale);
  const startTime = event.start_time ? new Date(event.start_time) : new Date();
  const endTime = event.end_time ? new Date(event.end_time) : null;

  const start: DateTime = getDateArray(startTime);
  const end: DateTime = endTime ? getDateArray(endTime) : start;
  const eventAttributes: EventAttributes = {
    start,
    end,
    title: name,
    description: shortDescription,
    location: getEventLocationText({ event, locale }),
  };

  return eventAttributes;
};

export const downloadEventIcsFile = async ({
  addNotification,
  event,
  locale,
  t,
}: {
  addNotification: AddNotificationFn;
  event: Event;
  locale: Language;
  t: TFunction;
}) => {
  try {
    const filename = `event_${event.id}.ics`;
    const file: File = await new Promise((resolve, reject) => {
      createEvent(getEventAttributes({ event, locale }), (error, value) => {
        if (error) {
          reject(error);
        }
        resolve(new File([value], filename, { type: 'text/calendar' }));
      });
    });

    const href = URL.createObjectURL(file);
    const anchor = document.createElement('a');
    anchor.download = filename;
    anchor.href = href;
    anchor.click();

    URL.revokeObjectURL(href);
  } catch {
    addNotification({
      label: t('common:eventCalendarButton.failedToCreateFile'),
      type: 'error',
    });
  }
};
