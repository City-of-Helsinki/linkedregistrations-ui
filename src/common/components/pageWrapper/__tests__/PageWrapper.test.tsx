import React from 'react';

import { configure, render } from '../../../../utils/testUtils';
import PageWrapper from '../PageWrapper';

configure({ defaultHidden: true });

const renderComponent = () =>
  render(<PageWrapper backgroundColor="coatOfArms" />);

test('should render component', () => {
  const { container } = renderComponent();

  expect(container).toMatchSnapshot();
});
