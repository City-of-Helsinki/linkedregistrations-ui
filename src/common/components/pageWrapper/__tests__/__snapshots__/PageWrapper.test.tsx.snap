// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
<div>
  <div
    class="pageWrapper backgroundColorCoatOfArms"
  />
</div>
`;
